src/main/scala/chisel3/util/Queue.scala

```diff
@@ -185,6 +187,37 @@ object Queue {
     }
   }
 
+  def withShadowLayer[T <: Data, S <: Data](
+    enq:            ReadyValidIO[T],
+    entries:        Int = 2,
+    pipe:           Boolean = false,
+    flow:           Boolean = false,
+    useSyncReadMem: Boolean = false,
+    flush:          Option[Bool] = None,
+    layerDataTuple: (S, layer.Layer)
+  ): (DecoupledIO[T], Valid[S]) = {
+    val queue = apply(enq, entries, pipe, flow, useSyncReadMem, flush)
```

Comment:
what if instead of immediately executing this thunk, it instead returned the thunk that could be executed in some far-away parent? using BoringUtils.tap to the signals instead of just ProbeValue?

Reply:
in that case it may be difficult to take the `S` directly, you might need to return the `S (in) and Valid[S] (out)` as well.

Reply:
This is a great idea. The types aren't problematic for the `module Queue`, but I need an intermediary class in order to make this work for the method commented on here. I can make this a `class ShadowFactory` which has a polymorphic apply method, I think.

But, ergonomically, what you're suggesting is fantastic. You create a queue and you get back a function that you can use to shadow it anywhere in the design. 💯 This also solves a problem that I was worried about which is needing to shadow the queue in multiple layers.

Reply:
Take a look at the latest version which includes a `ShadowFactory`. This can be used to construct lockstep shadow queues anywhere in the hierarchy.